### PR TITLE
add auto-formatting scripts and installable hook

### PR DESCRIPTION
**Pull request by rwcarlsen**
Adds some scripts for auto-formatting code:

* ``moose-format.sh``: basically an augmented CIVET pre-check script that uses ``clang-format`` instead of perl scripts (if available) to tell the user their code is "bad".
* ``autofmt.sh``: a separate script for auto-generating a patch of clang-format differences (and optionally applying it automatically).
* ``install-format-hook.sh``: installs a git pre-commit hook that prints out formatting violations (using ``moose-format.sh`` and ``autofmt.sh``).  This can easily be tweaked to auto-fix and include the changes in the commit, but I don't recommend that approach - you should integrate clang-format with your editor of choice instead.

This is in support of a fix for #7289.  We still need to also:

* get CIVET support for this in
* Update site/wiki with information about how to use these tools
* Update style guide
* convert appropriate parts of MOOSE to the new style.

But this doesn't break anything and can go in now - if we agree it looks reasonable.

### CODE REVIEW

**Comment by permcody** [comment]
Looking good, the only comment I have is it'd be nice to have the script read a list of directories to check. Right now those directories are encoded in a function buried inside of one of these bash scripts and it'll be a pain to expand that list as is. Can we just add it to the list of "config" variables at the top of `autofmt.sh`?

**Comment by rwcarlsen** [comment]
Added config var to autofmt.sh for directories list.

File: scripts/delete_trailing_whitespace.sh [hunk 1]
```diff
@@ -4,6 +4,8 @@
 # will be used (one up from the scripts directory where this script is located)
 REPO_DIR=${1:-"$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/../"}
 
+echo "this script is deprecated, please use 'moose-format.sh' instead"
```
Comment (brianmoose) [inline scripts/delete_trailing_whitespace.sh]: Don't we still need this script for `*.py`, `*.i`, and `*.tex` files?
And it seems like `moose-format.sh` only does checking...

**Comment by rwcarlsen** [comment]
ping

**Comment by dschwen** [comment]
Sounds good. I'll be adding https://github.com/LiquidHelium/atom-clang-format to mooseframework.org/wiki/atomio once this is merged.

**Review by permcody — APPROVED** [review]
Looks good!

**Pull request merged**